### PR TITLE
Fixing race condition

### LINKED ISSUE #184: Update value under concurrency fails when a cancellation token is registered.
### Scenario:
1. Set up a memory cache.
2. Add an entry with a cancellation token registered with it.
3. Start a child thread which keep reading the memory cache with the same key.
4. Reset the value of the entry from main thread.
5. Exception `InvalidOperationException` is thrown.

### Issue repo:
https://github.com/troydai/CachingIssues/blob/2a85107b9b572c47d66d2a0303deafecaeaa3166/TestActions/ConcurrencyIssue184.cs#L17

Run the sample follow exception will be thrown from child thread after the value is updated by main thread.

```
Unhandled Exception: System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
   at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
   at System.Collections.Generic.List`1.Enumerator.MoveNextRare()
   at Microsoft.Extensions.Caching.Memory.CacheEntry.PropageOptions(CacheEntry parent)
   at Microsoft.Extensions.Caching.Memory.MemoryCache.TryGetValue(Object key, Object& result)
   at Microsoft.Extensions.Caching.Memory.CacheExtensions.TryGetValue[TItem](IMemoryCache cache, Object key, TItem& value)
   at Microsoft.Extensions.Caching.Memory.CacheExtensions.Get[TItem](IMemoryCache cache, Object key)
   at ConsoleApplication.Program.<>c__DisplayClass5_0.<Monitor>b__0(Int32 round, DateTime beginning)
   at ConsoleApplication.Program.<>c__DisplayClass4_0.<Polling>b__0()
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
```

### PR DESCRIPTION
**Pull request by sebastienros**
Fixes #184

### CODE REVIEW

File: src/Microsoft.Extensions.Caching.Memory/CacheEntry.cs [hunk 1]
```diff
@@ -18,18 +18,19 @@ namespace Microsoft.Extensions.Caching.Memory
...
+        private IList<IDisposable> _expirationTokenRegistrations;
+        private EvictionReason _evictionReason;
         private IList<PostEvictionCallbackRegistration> _postEvictionCallbacks;
+        private bool _isExpired;
+
         internal IList<IChangeToken> _expirationTokens;
         internal DateTimeOffset? _absoluteExpiration;
         internal TimeSpan? _absoluteExpirationRelativeToNow;
         private TimeSpan? _slidingExpiration;
+        private IDisposable _bookmark;
```
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Memory/CacheEntry.cs]: needs a more descriptive name.

File: src/Microsoft.Extensions.Caching.Memory/CacheEntryHelper.cs [hunk 2]
```diff
@@ -49,31 +49,79 @@ internal static CacheEntry Current
...
+            var scopes = GetOrCreateScopes();
+
+            var bookmark = new CacheEntryStackBookmark(scopes);
+            Scopes = scopes.Push(entry);
+
+            return bookmark;
+        }
 
-                return null;
+        static CacheEntryStack GetOrCreateScopes()
```
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Memory/CacheEntryHelper.cs]: you're missing several access markers. Many of these should be private?

File: src/Microsoft.Extensions.Caching.Memory/CacheEntryHelper.cs [hunk 3]
```diff
@@ -49,31 +49,79 @@ internal static CacheEntry Current
...
 
-            Scopes.Push(entry);
+            public void Dispose()
+            {
+                Scopes = _bookmark;
+            }
+        }
+    }
+
+    class CacheEntryStack
```
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Memory/CacheEntryHelper.cs]: new file

File: test/Microsoft.Extensions.Caching.Memory.Tests/CacheEntryScopeExpirationTests.cs [hunk 4]
```diff
@@ -5,7 +5,9 @@
 using System.Linq;
 using Microsoft.Extensions.Internal;
 using Microsoft.Extensions.Caching.Memory.Infrastructure;
+using Microsoft.Extensions.Caching;
 using Xunit;
+using System.Threading.Tasks;
```
Comment (Tratcher) [inline test/Microsoft.Extensions.Caching.Memory.Tests/CacheEntryScopeExpirationTests.cs]: Sort

File: src/Microsoft.Extensions.Caching.Memory/CacheEntryHelper.cs [hunk 5]
```diff
@@ -49,31 +49,79 @@ internal static CacheEntry Current
...
+            {
+                scopes = CacheEntryStack.Empty;
+                Scopes = scopes;
             }
+
+            return scopes;
         }
 
-        internal static void EnterScope(CacheEntry entry)
+        sealed class CacheEntryStackBookmark : IDisposable
```
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Memory/CacheEntryHelper.cs]: CacheEntryStackMarker? CacheEntryStackEntry? CacheEntryStackLevel? StackEntry?

**Comment by sebastienros** [comment]
:up: :date: with some new name for the bookmark variable and class. Feel free to dislike it.

File: src/Microsoft.Extensions.Caching.Memory/CacheEntry.cs [hunk 6]
```diff
@@ -52,12 +53,12 @@ internal class CacheEntry : ICacheEntry
...
                 throw new ArgumentNullException(nameof(notifyCacheOfExpiration));
             }
-            
+
             Key = key;
             _notifyCacheEntryDisposed = notifyCacheEntryDisposed;
             _notifyCacheOfExpiration = notifyCacheOfExpiration;
 
-            CacheEntryHelper.EnterScope(this);
+            _scopes = CacheEntryHelper.EnterScope(this);
```
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Memory/CacheEntry.cs]: Better. Why not just `_scope`? There's only one of them as far as the CacheEntry is concerned.

**Comment by Tratcher** [comment]
:shipit:

**Pull request merged**